diff -u -r qsopcast-0.3.5.old/src/qsopcast.pro qsopcast-0.3.5/src/qsopcast.pro
--- qsopcast-0.3.5.old/src/qsopcast.pro	2008-11-08 18:54:29.000000000 +0200
+++ qsopcast-0.3.5/src/qsopcast.pro	2008-11-08 18:55:16.000000000 +0200
@@ -21,19 +21,6 @@
 MOC_DIR=.moc
 OBJECTS_DIR=.obj
 
-# Alsa
-ALIB=$$system(ls /usr/lib|grep libasound)
-contains(ALIB,libasound.so){
-  LIBS+=-lasound
-  DEFINES+=ALSA
-}else{
-  ALIB=$$system(ls /usr/local/lib|grep libasound)
-  contains(ALIB,libasound.so){
-    LIBS+=-lasound
-    DEFINES+=ALSA
-  }
-}
-
 
 # Input
 HEADERS += channel.h \
