--- data/table/Makefile.am.orig	2010-11-11 16:58:39.000000000 +0800
+++ data/table/Makefile.am	2010-11-11 16:58:39.000000000 +0800
@@ -1,8 +1,8 @@
 tabledir = $(pkgdatadir)/data/table
 
-mb_data = cj.mb erbi.mb wanfeng.mb wbpy.mb qxm.mb db.mb wbx.mb zrm.mb
-txt_data = cj.txt erbi.txt wanfeng.txt wbpy.txt qxm.txt db.txt wbx.txt zrm.txt
-mb_conf = qxm.conf db.conf erbi.conf wanfeng.conf  wbpy.conf wbx.conf zrm.conf cangjie.conf disable-table.conf.template
+mb_data = cj.mb erbi.mb wanfeng.mb wbpy.mb qxm.mb db.mb wbx.mb zrm.mb zhengma.mb
+txt_data = cj.txt erbi.txt wanfeng.txt wbpy.txt qxm.txt db.txt wbx.txt zrm.txt zhengma.txt
+mb_conf = qxm.conf db.conf erbi.conf wanfeng.conf  wbpy.conf wbx.conf zrm.conf zhengma.conf cangjie.conf disable-table.conf.template
 
 table_DATA = $(mb_data) $(mb_conf)
 EXTRA_DIST = $(mb_conf) $(txt_data)
