diff -Naur mldonkey-3.0.0-orig/config/configure.in mldonkey-3.0.0-ptch/config/configure.in
--- mldonkey-3.0.0-orig/config/configure.in	2009-02-24 21:33:03.000000000 +0000
+++ mldonkey-3.0.0-ptch/config/configure.in	2009-11-02 22:33:31.000000000 +0000
@@ -486,7 +486,7 @@
 else
   OCAMLVERSION=`$OCAMLC -v | sed -n -e 's|.*version* *\(.*\)$|\1|p' `
   case "$OCAMLVERSION" in
-    "$REQUIRED_OCAML"*|3.11.0*|3.10.*|3.09.*|3.08.4*|3.08.3*) ;;
+    "$REQUIRED_OCAML"*|3.11.*|3.10.*|3.09.*|3.08.4*|3.08.3*) ;;
     *)
         echo "Need build"
         BUILD_OCAML=yes
